<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc" style="margin-top: 1em;"><ul class="toc-item"><li><span><a href="#Command-line-tools-for-wrangling-Twitter-data" data-toc-modified-id="Command-line-tools-for-wrangling-Twitter-data-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Command-line tools for wrangling Twitter data</a></span><ul class="toc-item"><li><span><a href="#wc" data-toc-modified-id="wc-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>wc</a></span><ul class="toc-item"><li><span><a href="#Count-the-tweets" data-toc-modified-id="Count-the-tweets-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>Count the tweets</a></span></li></ul></li><li><span><a href="#gzip" data-toc-modified-id="gzip-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>gzip</a></span><ul class="toc-item"><li><span><a href="#Compress-a-stream-of-tweets" data-toc-modified-id="Compress-a-stream-of-tweets-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Compress a stream of tweets</a></span></li><li><span><a href="#Uncompress-to-a-stream-of-tweets" data-toc-modified-id="Uncompress-to-a-stream-of-tweets-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Uncompress to a stream of tweets</a></span></li></ul></li><li><span><a href="#awk" data-toc-modified-id="awk-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>awk</a></span><ul class="toc-item"><li><span><a href="#Sample-tweets" data-toc-modified-id="Sample-tweets-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Sample tweets</a></span></li></ul></li><li><span><a href="#split" data-toc-modified-id="split-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>split</a></span><ul class="toc-item"><li><span><a href="#Split-tweets-into-files-by-number-of-tweets" data-toc-modified-id="Split-tweets-into-files-by-number-of-tweets-1.4.1"><span class="toc-item-num">1.4.1&nbsp;&nbsp;</span>Split tweets into files by number of tweets</a></span></li></ul></li><li><span><a href="#jq" data-toc-modified-id="jq-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>jq</a></span><ul class="toc-item"><li><span><a href="#Extract-mentions" data-toc-modified-id="Extract-mentions-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>Extract mentions</a></span></li></ul></li><li><span><a href="#sort" data-toc-modified-id="sort-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>sort</a></span><ul class="toc-item"><li><span><a href="#Sort-a-list" data-toc-modified-id="Sort-a-list-1.6.1"><span class="toc-item-num">1.6.1&nbsp;&nbsp;</span>Sort a list</a></span></li></ul></li><li><span><a href="#uniq" data-toc-modified-id="uniq-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>uniq</a></span><ul class="toc-item"><li><span><a href="#Get-the-unique-values-for-a-list" data-toc-modified-id="Get-the-unique-values-for-a-list-1.7.1"><span class="toc-item-num">1.7.1&nbsp;&nbsp;</span>Get the unique values for a list</a></span></li></ul></li><li><span><a href="#tee" data-toc-modified-id="tee-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>tee</a></span><ul class="toc-item"><li><span><a href="#Read-once,-write-twice" data-toc-modified-id="Read-once,-write-twice-1.8.1"><span class="toc-item-num">1.8.1&nbsp;&nbsp;</span>Read once, write twice</a></span></li></ul></li><li><span><a href="#parallel" data-toc-modified-id="parallel-1.9"><span class="toc-item-num">1.9&nbsp;&nbsp;</span>parallel</a></span><ul class="toc-item"><li><span><a href="#gzip-in-parallel" data-toc-modified-id="gzip-in-parallel-1.9.1"><span class="toc-item-num">1.9.1&nbsp;&nbsp;</span>gzip in parallel</a></span></li></ul></li><li><span><a href="#json2csv.py" data-toc-modified-id="json2csv.py-1.10"><span class="toc-item-num">1.10&nbsp;&nbsp;</span>json2csv.py</a></span><ul class="toc-item"><li><span><a href="#Convert-JSON-tweets-to-CSV" data-toc-modified-id="Convert-JSON-tweets-to-CSV-1.10.1"><span class="toc-item-num">1.10.1&nbsp;&nbsp;</span>Convert JSON tweets to CSV</a></span></li></ul></li><li><span><a href="#csvkit" data-toc-modified-id="csvkit-1.11"><span class="toc-item-num">1.11&nbsp;&nbsp;</span>csvkit</a></span><ul class="toc-item"><li><span><a href="#Select-columns" data-toc-modified-id="Select-columns-1.11.1"><span class="toc-item-num">1.11.1&nbsp;&nbsp;</span>Select columns</a></span></li><li><span><a href="#Merge-CSVs" data-toc-modified-id="Merge-CSVs-1.11.2"><span class="toc-item-num">1.11.2&nbsp;&nbsp;</span>Merge CSVs</a></span></li><li><span><a href="#Filter-CSVs" data-toc-modified-id="Filter-CSVs-1.11.3"><span class="toc-item-num">1.11.3&nbsp;&nbsp;</span>Filter CSVs</a></span></li><li><span><a href="#Join-reply-tweets-and-the-tweets-that-are-replied-to" data-toc-modified-id="Join-reply-tweets-and-the-tweets-that-are-replied-to-1.11.4"><span class="toc-item-num">1.11.4&nbsp;&nbsp;</span>Join reply tweets and the tweets that are replied to</a></span></li></ul></li><li><span><a href="#twarc-utilities" data-toc-modified-id="twarc-utilities-1.12"><span class="toc-item-num">1.12&nbsp;&nbsp;</span>twarc utilities</a></span><ul class="toc-item"><li><span><a href="#tweet_compliance.py" data-toc-modified-id="tweet_compliance.py-1.12.1"><span class="toc-item-num">1.12.1&nbsp;&nbsp;</span>tweet_compliance.py</a></span></li><li><span><a href="#deduplicate.py" data-toc-modified-id="deduplicate.py-1.12.2"><span class="toc-item-num">1.12.2&nbsp;&nbsp;</span>deduplicate.py</a></span></li><li><span><a href="#deletes.py" data-toc-modified-id="deletes.py-1.12.3"><span class="toc-item-num">1.12.3&nbsp;&nbsp;</span>deletes.py</a></span></li><li><span><a href="#unshrtn.py" data-toc-modified-id="unshrtn.py-1.12.4"><span class="toc-item-num">1.12.4&nbsp;&nbsp;</span>unshrtn.py</a></span></li></ul></li></ul></li></ul></div>

# Command-line tools for wrangling Twitter data
Sure, you could write a script but you can often get the job done from the command-line.

This is an assortment of command-line tools that I use for wrangling Twitter data. Of course, most of these tools and techniques can be used for wrangle other types of data. If you have others, please let me know.

To illustrate the tools, I retrieved the tweets posted by [@gelmanlibrary](https://twitter.com/gelmanlibrary) and [@gwtweets](https://twitter.com/gwtweets) using DocNow's [Twarc](https://github.com/docnow/twarc). Twarc is a command-line tool for retrieving data from Twitter's API.

    twarc timeline gwtweets > gwtweets.jsonl
    twarc timeline gelmanlibrary > gelmanlibrary.jsonl
    
`gwtweets.jsonl` and `gelmanlibrary.jsonl` are line-oriented JSON files, i.e., tweets are in JSON, with one tweet on each line.

*Mac tip: To get Linux-like command `brew install coreutils`*

## wc
### Count the tweets

In [1]:
!wc -l *.jsonl

    3223 gelmanlibrary.jsonl
    3218 gwtweets.jsonl
    6441 total


*wc gotcha: When counting many tweets, wc -l will output a partial total and then reset the count.*

## gzip
Because of the size of the datasets that I deal with, I usually gzip compress the tweets.
### Compress a stream of tweets

In [2]:
!cat gwtweets.jsonl | gzip > gwtweets.jsonl.gz
!cat gelmanlibrary.jsonl | gzip > gelmanlibrary.jsonl.gz
!ls *.jsonl.gz

gelmanlibrary.jsonl.gz	gwtweets.jsonl.gz


### Uncompress to a stream of tweets

In [3]:
!gunzip -c *.jsonl.gz | wc -l

6441


## awk
### Sample tweets
In the following example, I create a 1 in 5 sample.

In [4]:
!gunzip -c *.jsonl.gz | awk 'NR % 5 == 0' > sample.json
!wc -l sample.json

1288 sample.json


## split
### Split tweets into files by number of tweets

In [5]:
!gunzip -c *.jsonl.gz | split --lines=1000 -d --additional-suffix=.jsonl - tweets-
!wc -l tweets-*.jsonl

    1000 tweets-00.jsonl
    1000 tweets-01.jsonl
    1000 tweets-02.jsonl
    1000 tweets-03.jsonl
    1000 tweets-04.jsonl
    1000 tweets-05.jsonl
     441 tweets-06.jsonl
    6441 total


## jq
[jq](https://stedolan.github.io/jq/) excels at transforming JSON data. Because jq is such a useful tool for Twitter data, we already have several blog posts dedicated to it:
* [Getting Started Working with Twitter Data Using jq](http://nbviewer.jupyter.org/github/gwu-libraries/notebooks/blob/master/20160407-twitter-analysis-with-jq/Working-with-twitter-using-jq.ipynb)
* [Recipes for processing Twitter data with jq](http://nbviewer.jupyter.org/github/gwu-libraries/notebooks/blob/master/20161122-twitter-jq-recipes/twitter_jq_recipes.ipynb)

However, here is an example of one of its uses.

### Extract mentions

In [6]:
!gunzip -c *.jsonl.gz | jq -r '.entities.user_mentions[].screen_name' > screen_names.txt
!head -10 screen_names.txt

GWAcadTech
Azodiac83
gelmanlibrary
GWtweets
GWPeterK
GWUSpecColl
GWTextileMuseum
gelmanlibrary
GWAcadTech
gelmanlibrary


## sort
### Sort a list
`-r` reverses the sort order (see the example for uniq).

In [7]:
!gunzip -c *.jsonl.gz | jq -r '.entities.user_mentions[].screen_name' | sort > sorted_screen_names.txt
!head -10 sorted_screen_names.txt

10p
3PM
60Minutes
60Minutes
60Minutes
60Minutes
60Minutes
6pm
8
8nKearns


## uniq
### Get the unique values for a list
The list must be sorted first. Omit `-c` to remove counts.

In [8]:
!gunzip -c *.jsonl.gz | jq -r '.entities.user_mentions[].screen_name' | sort | uniq -c | sort -r > unique_screen_names.txt
!head -10 unique_screen_names.txt

    881 gelmanlibrary
    241 GWPeterK
    240 EcklesLibrary
    133 GWtweets
    110 StudentLiaison
    104 GW_MBB
     82 GWDivIT
     81 GWToday
     78 stemworksgw
     75 GW_WBB


## tee
### Read once, write twice
In this example, I write out a lists of tweet ids and user ids.

In [9]:
%%bash
gunzip -c gwtweets.jsonl.gz | tee >(jq -r '.user.id_str' > gwtweets-user_ids.txt) | jq -r '.id_str' > gwtweets-tweet_ids.txt
head -5 gwtweets-tweet_ids.txt
head -5 gwtweets-user_ids.txt

976161920133816322
976147556987297794
976144727610454016
976141006365253632
976139936696020993
28101965
28101965
28101965
28101965
28101965


## parallel
parallel can really speed up processes that involve multiple files. It is also useful for repeating a task multiple times, substituting in values listed in a file.

`-j` controls the number of parallel processes. You should choose an appropriate number for the number of free CPUs available.

### gzip in parallel

In [10]:
%%bash
ls -1 tweets-*.jsonl > src.lst
cat src.lst | sed 's/.jsonl/.jsonl.gz/' > dest.lst
parallel -a src.lst -a dest.lst -j 2 --xapply "cat {1} | gzip > {2}"
ls tweets-*.jsonl.gz

tweets-00.jsonl.gz
tweets-01.jsonl.gz
tweets-02.jsonl.gz
tweets-03.jsonl.gz
tweets-04.jsonl.gz
tweets-05.jsonl.gz
tweets-06.jsonl.gz


## json2csv.py
json2csv.py is a utility that is part of Twarc.

In [11]:
!git clone https://github.com/DocNow/twarc.git
!pip install -e twarc

Cloning into 'twarc'...
remote: Counting objects: 2606, done.
remote: Total 2606 (delta 0), reused 0 (delta 0), pack-reused 2606
Receiving objects: 100% (2606/2606), 659.15 KiB | 3.42 MiB/s, done.
Resolving deltas: 100% (1648/1648), done.
Obtaining file:///Users/justinlittman/Data/notebooks/20180320-twitter-commandline/twarc
  Found existing installation: twarc 1.3.9
    Uninstalling twarc-1.3.9:
      Successfully uninstalled twarc-1.3.9
  Running setup.py develop for twarc


### Convert JSON tweets to CSV
The `-x` option produces Excel friendly CSV by removing newlines from text fields such as the tweet.

In [12]:
!gunzip -c gwtweets.jsonl.gz | python twarc/utils/json2csv.py -x - > gwtweets.csv
!gunzip -c gelmanlibrary.jsonl.gz | python twarc/utils/json2csv.py -x - > gelmanlibrary.csv
!head -2 gwtweets.csv

*Tip when loading a tweet CSV into Excel*: If you open up a tweet CSV with Excel, it will mishandle tweet and user ids. For example, 976161920133816322 will become 976161920133816000.

To correctly import tweet CSV into Excel, select Data > Get External Data > Import File. When given the option of selecting the data type for fields, select text for all id fields.

## csvkit
[csvkit](http://csvkit.readthedocs.io) supports a wide variety of operations for filtering and transforming CSV files. Here are a few highlights.

In [13]:
!pip install csvkit

### Select columns

In [14]:
!csvcut -c id,created_at,text gelmanlibrary.csv > gelmanlibrary_cut.csv
!head -3 gelmanlibrary_cut.csv

id,created_at,text
974067184249995265,Wed Mar 14 23:38:08 +0000 2018,"RT @GWAcadTech: The Instructional Technology Lab is available for support, including appointments and walk-in hours https://t.co/veLDwGcbWo…"
973974487384363008,Wed Mar 14 17:29:48 +0000 2018,SO MANY #GWU people here to learn about GW Box today!   Attend one of our free productivity tools workshops this week at 1pm. Thu: Getting More Out Of Spreadsheets Fri: Google Forms https://t.co/hk8MfUMeit  All the cool kids are doing it. https://t.co/Gsiib3ZSZ5


### Merge CSVs

In [15]:
!csvstack gelmanlibrary.csv gwtweets.csv > merged.csv
!wc -l *.csv

   3224 gelmanlibrary.csv
   3224 gelmanlibrary_cut.csv
     82 gelmanlibrary_replies_to.csv
   3219 gwtweets.csv
   6442 merged.csv
  16191 total


### Filter CSVs
Here I'm finding all of the tweets that are replies and getting the tweets to which they are replies.

In [16]:
!csvgrep -c tweet_type -m reply gelmanlibrary.csv | csvcut -c id,in_reply_to_status_id > replies_to_in_reply_to.csv
!head -5 replies_to_in_reply_to.csv

id,in_reply_to_status_id
970826158907514880,970825298471215105
960266436827602948,960265847108526081
938232259173322752,938231696012562432
938136395352289281,938135859211186176


### Join reply tweets and the tweets that are replied to
Here I'm joining some @gelmanlibrary replies to the tweets that they are a reply to.

The @gelmanlibrary reply will be on the left; the tweet that is being replied to will be on the right (with field names appended with "2").

Be careful not to use csvjoin on large CSVs.

In [17]:
!csvjoin -c in_reply_to_status_id,id gelmanlibrary.csv gelmanlibrary_replies_to.csv > gelmanlibrary_with_replies_to.csv
!head -2 gelmanlibrary_with_replies_to.csv

id,tweet_url,created_at,parsed_created_at,user_screen_name,text,tweet_type,coordinates,hashtags,media,urls,favorite_count,in_reply_to_screen_name,in_reply_to_status_id,in_reply_to_user_id,lang,place,possibly_sensitive,retweet_count,reweet_or_quote_id,retweet_or_quote_screen_name,retweet_or_quote_user_id,source,user_id,user_created_at,user_default_profile_image,user_description,user_favourites_count,user_followers_count,user_friends_count,user_listed_count,user_location,user_name,user_statuses_count,user_time_zone,user_urls,user_verified,tweet_url2,created_at2,parsed_created_at2,user_screen_name2,text2,tweet_type2,coordinates2,hashtags2,media2,urls2,favorite_count2,in_reply_to_screen_name2,in_reply_to_status_id2,in_reply_to_user_id2,lang2,place2,possibly_sensitive2,retweet_count2,reweet_or_quote_id2,retweet_or_quote_screen_name2,retweet_or_quote_user_id2,source2,user_id2,user_created_at2,user_default_profile_image2,user_description2,user_favourites_count2,user_followers_count2,user_frie

## twarc utilities
In addition to json2csv.py, Twarc includes a number of other useful tweet utilities ([docs](https://github.com/docnow/twarc#utilities) and [complete list of scripts](https://github.com/DocNow/twarc/tree/master/utils)).

Here are some of my favorites.

### tweet_compliance.py
Supports [tweet compliance](https://developer.twitter.com/en/docs/tweets/compliance/overview) by retrieving the most current versions of tweets or removing unavailable (deleted or protected)
tweets.

Also useful for splitting out deleted tweets.

### deduplicate.py
Not surpringly, deduplicates tweets. For a retweet, `--extract-retweets` will return the retweet and source tweet (i.e., the tweet that is retweeted). This is useful for extracting all of the tweets in a dataset.

### deletes.py
Attempts to determine why a tweet was deleted, e.g., tweet deleted, user protected, retweet deleted.

### unshrtn.py
Unshortens URLs contained in tweets and adds them to the tweet.

